Extract CPU log_softmax kernels to header

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156243

This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

swolchok committed:
Extract CPU log_softmax kernels to header

This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

[ghstack-poisoned]

janeyx99 commented on aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:
```diff
@@ -0,0 +1,325 @@
+    int64_t d0 = 0;
+    for (; d0 < size - (size % Vec::size()); d0 += Vec::size()) {
+      min_fvec.store(input_max_data + d0);
+      min_fvec.store(input_max_data + d0 + fVec::size());
```
This looks added from the original implementation (and I see this throughout the below) -- are we doing 2x as much vectorization? Or...?

swolchok replied:
it's not new; there are multiple kernels being moved (there's an enable_if). you can verify moved code in Phabricator

janeyx99 commented on aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:
```diff
@@ -0,0 +1,325 @@
+
+    // compute sum of (x - max).exp()
+    for (int64_t dim_idx = 0; dim_idx < dim_size; dim_idx++) {
+      float* input_buffer_ptr = input_buffer_data + dim_idx * chunk_size;
```
how come float* instead of const scalar_t*?

swolchok replied:
not new

janeyx99 reviewed (commented):
CI will check correctness--is there a plan to check perf?

swolchok committed:
add missing functional.h include on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

janeyx99 reviewed (approved):
Oooh

swolchok commented:
@pytorchbot merge

swolchok commented:
hmm, these linux-aarch64 failures look nonsensical (test_forward_nn_CrossEntropyLoss_cpu_float64 is failing). rebasing.

swolchok commented:
hm, second failure does mention log_softmax, perhaps it is related

swolchok committed:
fix a couple minor issues in case those caused test failures on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

swolchok committed:
rebase / kick CI so I can debug linux-aarch64 failures on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

swolchok committed:
debugging change on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

swolchok committed:
remove unused grain_size parameter to vec_logsoftmax_chunk_size_and_num_chunks on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

swolchok committed:
forgot to save the header on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

janeyx99 commented on aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:
```diff
@@ -0,0 +1,328 @@
+  int64_t MAX_CHUNK_SIZE = std::max<int64_t>(BLOCK_SIZE / dim_size / sizeof(scalar_t), Vec::size());
+  MAX_CHUNK_SIZE = MAX_CHUNK_SIZE / Vec::size() * Vec::size();
+  int64_t CHUNK_SIZE = std::min<int64_t>(MAX_CHUNK_SIZE, inner_size);
+  int64_t num_chunks = (inner_size + CHUNK_SIZE - 1) / CHUNK_SIZE;
```
perf optimization instead of using divup?

swolchok replied:
I can't use divup because it's in ATen/Parallel.h or ATen/Parallel-inl.h, which aren't standalone.

malfet replied:
Can you just re-implement `div_up` in the header (as static inline) and add todo to merge it?

swolchok commented on aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:
```diff
@@ -0,0 +1,328 @@
+  if (end <= begin) {
+    return;
+  }
+  using Vec = vec::Vectorized<vec::vec_scalar_t<scalar_t>>;
```
This is meant to be coming from the old line 39 of SoftMaxKernel.cpp, which is the first line of _vec_log_softmax_lastdim.

swolchok commented:
failures here are hardware-specific; tests are failing on graviton 4 (m8g) but not graviton 3 (m7g)

swolchok committed:
kick CI so I can try SSH+debug again on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

swolchok commented:
Huge props to @malfet for solving this one. I forgot `inline namespace CPU_CAPABILITY` in my extracted header (it wasn't in the source file because everything was in an anonymous namespace). m8g instances (Graviton 4) have 128-bit SVE support and we only support SVE256. What happened was the SVE256 code incorrectly ran anyway on m8g, which of course silently ran it with 128-bit vector width, hence output item number 5 being uninitialized.

swolchok committed:
add forgotten inline namespace CPU_CAPABILITY on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

swolchok committed:
Update on "Extract CPU log_softmax kernels to header"


This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge
